chain-storage/src/lib.rs

```diff
@@ -1,1079 +1,809 @@
+//! | Block 1 |
+//! |         |
+//! +---------+
+//!      |
+//!      |
+//!      v
+//! +---------+
+//! |         |
+//! |  (root) |
+//! |         |
+//! +---------+
+//! ```
+//!
+//! ## Removing stale branches
+//!
+//! If you want to clean up branches, do the following:
+//!
+//! * Call `store.get_tips_ids()`, in our example it will return
+//!   `[Block 4 id, Block 4' id]`;
+//! * Determine which branches do you want to remove.
+//! * Call, for example, `store.prune_branch(Block 4' id)`.
+//!
+//! ## Performance benefits of permanent storage
+//!
+//! Since blocks in the permanent storage are stored just one after another (the
+//! structure is `block_length.to_le_bytes() ++ block_bytes` and a file with
+//! references to blocks in the order they were added to the storage), it allows
+//! for the following scenarios:
+//!
+//! * Very fast block iteration
+//! * Transferring a portion of the data file over the network without locking
+//!   it.
+//! * O(1) access by chain length.
+//!
+//! __fig. 2 - permanent storage structure__
+//!
+//! ```ignore
+//! store                  block no. index
+//!
+//! +--------------+       +-------------+
+//! |              |       |             |
+//! | block1       |       | block1 pos  |
+//! |              |       |             |
+//! +--------------+       +-------------+
+//! |              |       |             |
+//! | block2       |       | block2 pos  |
+//! |              |       |             |
+//! +--------------+       +-------------+
+//! |              |       |             |
+//! | ...          |       | ...         |
+//! |              |       |             |
+//! +--------------+       +-------------+
+//! |              |       |             |
+//! | blockn       |       | blockn pos  |
+//! |              |       |             |
+//! +--------------+       +-------------+
+//! ```
+//!
+//! # Storage directory layout
+//!
+//! ```ignore
+//! store
+//! ├── permanent       - permanent storage directory
+//! │   └── flatfile    - storage file that can be transferred over the network
+//! └── volatile        - volatile storage
+//! ```
+
+mod block_info;
+mod error;
+mod iterator;
+mod permanent_store;
+#[cfg(any(test, feature = "with-bench"))]
+pub mod test_utils;
+#[cfg(test)]
+mod tests;
+mod value;
+
+use permanent_store::PermanentStore;
+use sled::{ConflictableTransactionError, TransactionError, Transactional, TransactionalTree};
+use std::path::Path;
+
+pub use block_info::BlockInfo;
+pub use error::Error;
+pub use iterator::StorageIterator;
+pub use value::Value;
+
+#[derive(Clone)]
+pub struct BlockStore {
+    volatile: sled::Db,
+    permanent: PermanentStore,
+    root_id: Value,
+    id_length: usize,
 }
 
-#[derive(Clone, Debug)]
-pub struct BackLink<Id: BlockId> {
-    /// The distance to this ancestor.
-    pub distance: u64,
-    /// The hash of the ancestor.
-    pub block_hash: Id,
+enum RemoveTipResult {
+    NextTip(Vec<u8>),
+    HitPermanentStore(Vec<u8>),
```

Comment:
OK, I've learned that block IDs are just bags of bytes just like every other values for storage.

Something like this would be more self-explanatory:

```rust
enum RemoveTipResult {
    NextTip { id: Vec<u8> },
    HitPermanentStore { id: Vec<u8> },
    Done,
}
```